### PR TITLE
Webhook event queue: debounce per-PR timer

### LINKED ISSUE #389: Webhook event queue: reset delay counter on each new event to prevent premature delivery
## Problem

When multiple GitHub webhook events arrive in quick succession (e.g. several check_run completions + a review comment), the foreman can receive and act on an earlier event before all related events have been delivered. The delay counter should be reset each time a new event arrives in the same burst, so the foreman only processes the batch once the stream has gone quiet.

## Expected behaviour

- On each incoming webhook event, reset a debounce timer (e.g. 2–5 seconds).
- Only deliver / surface the event(s) to the foreman once no new events have arrived within that window.
- This mirrors a standard debounce pattern and prevents the foreman from reacting to a half-delivered burst.

## Possible approach

In the webhook ingestion layer, maintain a per-PR (or per-task) timer. Each new event for the same PR resets the timer. Fire the foreman callback only on timer expiry.

### PR DESCRIPTION
**Pull request by jmelloy**
## Summary

- Adds a per-task (per-PR) debounce timer in `backend/routes/webhooks.py` with a 3-second window (`DEBOUNCE_WINDOW_SECONDS`).
- Each new dispatchable webhook event for the same PR cancels the previous timer and restarts the window. Only when no further events arrive within the window does the foreman get invoked.
- All summaries buffered during the window are joined with `---` separators and delivered together, so the foreman sees the complete picture of what happened.
- Updates two existing dispatch tests to patch `_schedule_debounced_foreman` (the new entry point) instead of `spawn`/`reset_foreman_poll` directly.
- Adds `TestDebounce` with four async tests: rapid-events coalescing → single delivery; slow events → separate deliveries; timer reset cancels prior timer; independent PR keys have independent timers.

## Test plan

- [x] All 412 backend tests pass (`python -m pytest`)
- [x] `ruff check . && ruff format .` — no issues
- [x] `TestDebounce::test_rapid_events_single_foreman_call` — 3 events → 1 foreman call with combined summary
- [x] `TestDebounce::test_slow_events_separate_foreman_calls` — events > window apart → 2 foreman calls
- [x] `TestDebounce::test_reset_cancels_previous_timer` — 2nd event before timer fires → combined in single delivery
- [x] `TestDebounce::test_independent_prs_not_merged` — events on different PR keys don't coalesce

Closes #389

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: backend/tests/test_github_webhooks_phase2.py [hunk 1]
```diff
@@ -349,15 +347,113 @@ def test_webhook_dispatches_for_ci_bot_check_run(client):
...
+        async def fake_run_foreman(guild_id, summary, *, user_id=None):
+            self._foreman_calls.append((guild_id, summary, user_id))
+
+        return (
+            patch.object(wh, "DEBOUNCE_WINDOW_SECONDS", 0.05),
+            patch.object(wh, "run_foreman_ai", new=fake_run_foreman),
+            patch.object(wh, "reset_foreman_poll"),
+            patch.object(wh, "spawn", side_effect=lambda coro, **kw: asyncio.create_task(coro)),
+        )
+
```
Comment (jmelloy) [inline backend/tests/test_github_webhooks_phase2.py:395]: These `async def` test methods have no `@pytest.mark.asyncio` decorator and `TestDebounce` has no asyncio mode configured. Without it (or a class-level `pytestmark`), pytest will collect them as coroutine objects, immediately return without awaiting, and mark them as passing — meaning the assertions never actually run. Add `@pytest.mark.asyncio` to each method or set `asyncio_mode = 'auto'` in the pytest config.

File: backend/routes/webhooks.py [hunk 2]
```diff
@@ -48,6 +61,64 @@
...
+    return f"{guild_id}:{task_id}"
+
+
+async def _debounce_fire(key: str, guild_id: str) -> None:
+    """Sleep for the debounce window then deliver all buffered events.
+
+    If this task is cancelled before the window expires (because a new event
+    arrived and reset the timer), the CancelledError propagates normally and
+    nothing is delivered — the replacement timer carries the full buffer.
+    """
```
Comment (jmelloy) [inline backend/routes/webhooks.py]: When a debounce task is cancelled (because a newer event arrived), `_debounce_fire` exits via `CancelledError` before calling `_debounce_buffers.pop`. The buffer entry is therefore left in `_debounce_buffers` until the replacement timer fires — this is intentional and correct. However, if the event loop shuts down while a task is in-flight (e.g. during testing or server shutdown), the buffer will never be drained and the accumulated entries will be silently discarded with no log. Consider adding a `try/finally` block that at minimum logs the abandoned buffer.

File: backend/routes/webhooks.py [hunk 3]
```diff
@@ -35,6 +36,18 @@
 
 router = APIRouter()
 
+# How long (seconds) to wait for further events on the same PR before
+# delivering the buffered batch to the foreman.  GitHub often sends several
+# check_run completions and a review comment within a second of each other;
+# this window lets them coalesce into a single foreman invocation.
+DEBOUNCE_WINDOW_SECONDS: float = 3.0
```
Comment (jmelloy) [inline backend/routes/webhooks.py]: These two module-level dicts are global singletons shared across all requests and tests. There is no cleanup mechanism (no TTL, no eviction on PR close, no shutdown hook). Long-running servers processing many PRs will accumulate keys indefinitely. Consider a `weakref`-based structure, explicit cleanup on PR close events, or at minimum a comment noting this limitation.

File: backend/routes/webhooks.py [hunk 4]
```diff
@@ -48,6 +61,64 @@
...
+    summary: str,
+    user_id: str | None,
+) -> None:
+    """Buffer *summary* and (re)start the per-PR debounce timer.
+
+    Each call resets the window so the foreman is only invoked after
+    DEBOUNCE_WINDOW_SECONDS of silence on this PR.
+    """
+    existing = _debounce_tasks.pop(key, None)
+    if existing and not existing.done():
```
Comment (jmelloy) [inline backend/routes/webhooks.py]: `user_id` is set to the last event's `user_id` (`items[-1][1]`). If a bot-triggered check_run completes after a human review comment within the debounce window, the foreman will be attributed to `None` or a bot user rather than the human. Consider using the first non-None user_id or the earliest event's user instead.

**Review by jmelloy — COMMENTED** [review]
- Debounce logic is sound and the `asyncio.Task` cancellation approach is correct, but the module-level mutable dictionaries (`_debounce_buffers`, `_debounce_tasks`) are global singletons that will bleed state between tests and across server restarts without an explicit cleanup path.
- The `_debounce_fire` function silently swallows the buffer if the task is cancelled mid-sleep (CancelledError propagates), but the buffer is never cleaned up in that case — new events append to it correctly, yet a cancelled task leaves stale entries until a new event arrives or the next fire.
- `user_id` selection strategy (always taking the last event's `user_id`) is implicit and undocumented; if the first event has a meaningful user and the last is a bot event, the foreman attribution will be wrong.
- The `TestDebounce` tests are `async def` methods on a plain class (no `pytest-asyncio` marker or `asyncio.run` wrapper), which means they will silently pass as no-ops in most default pytest configurations — they need `@pytest.mark.asyncio` or an equivalent mechanism.
- The `_patched_env` helper returns a plain tuple of context managers that must be unpacked manually (`p1, p2, p3, p4`), which is fragile; using `contextlib.ExitStack` or a single `@contextmanager` would be safer and more maintainable.

**Review by jmelloy — COMMENTED** [review]
- The debounce implementation is well-structured with a clean `DebounceQueue` class, proper shutdown handling, and good test coverage including edge cases like timer cancellation and state isolation.
- A critical race condition exists in `_fire`: the comment says the task is replaced *synchronously before* CancelledError is injected, but asyncio task cancellation is not synchronous — the CancelledError is delivered on the next event-loop iteration, meaning `self._tasks.get(key) is asyncio.current_task()` may be true even when a replacement timer exists, causing premature delivery of buffered events.
- The `test_cancelled_events_not_lost` test asserts that buffers are preserved after external cancellation, but the `_fire` method's CancelledError handler calls `self._buffers.pop(key, [])` when it detects it's still the registered task — which contradicts the test's expectation and reveals a logic inconsistency.
- The default debounce window of 60 seconds is very aggressive for a webhook handler; this could delay foreman invocations by up to a minute for legitimate single-event scenarios (e.g., a PR review with no follow-up), and the PR description mentions 3 seconds but the code defaults to 60.
- The `_deliver` method's user_id selection strategy (last event's user_id) is reasonable but undocumented in user-facing terms; also, silently dropping attribution from intermediate events with different user_ids could be a concern for audit trails.

File: backend/routes/webhooks.py [hunk 5]
```diff
@@ -48,6 +57,128 @@
...
+            task.cancel()
+        if pending:
+            await asyncio.gather(*pending, return_exceptions=True)
+        self._tasks.clear()
+        self._buffers.clear()
+
+    def _deliver(self, key: str, guild_id: str, items: list[tuple[str, str | None]]) -> None:
+        summaries = [s for s, _ in items]
+        combined = "\n\n---\n\n".join(summaries) if len(summaries) > 1 else summaries[0]
+        # Use the last event's user_id: in typical bursts the final event is a
```
Comment (jmelloy) [inline backend/routes/webhooks.py:105]: This race condition check is unreliable. `schedule()` cancels the old task and assigns a new one to `self._tasks[key]`, but the CancelledError is not injected until the next event-loop tick. By the time this handler runs, `self._tasks[key]` has already been replaced with the new task, so `self._tasks.get(key) is asyncio.current_task()` will be `False` — the early-delivery branch is unreachable in the normal replacement case. The comment's claim about synchronous replacement is incorrect for asyncio task cancellation.

File: backend/routes/webhooks.py [hunk 6]
```diff
@@ -35,6 +37,13 @@
 
 router = APIRouter()
 
+# How long (seconds) to wait for further events on the same PR before
+# delivering the buffered batch to the foreman.  GitHub often sends several
+# check_run completions and a review comment within a second of each other;
+# this window lets them coalesce into a single foreman invocation.
+# Configurable via WEBHOOK_DEBOUNCE_SECONDS environment variable (default: 60).
```
Comment (jmelloy) [inline backend/routes/webhooks.py]: The PR description says the debounce window is 3 seconds, but the default here is 60 seconds. A 60-second delay before foreman invocation is a significant UX regression for single-event webhooks (e.g., a lone PR review). Consider aligning with the PR description or explicitly justifying the 60-second default.

File: backend/tests/test_github_webhooks_phase2.py [hunk 7]
```diff
@@ -349,15 +348,142 @@ def test_webhook_dispatches_for_ci_bot_check_run(client):
...
+            wh._debounce_queue.schedule("g-ind:t-pr1", "g-ind", "pr1 event", "u4")
+            wh._debounce_queue.schedule("g-ind:t-pr2", "g-ind", "pr2 event", "u4")
+            await asyncio.sleep(0.2)
+
+        assert len(self._foreman_calls) == 2
+        summaries = {c[1] for c in self._foreman_calls}
+        assert any("pr1 event" in s for s in summaries)
+        assert any("pr2 event" in s for s in summaries)
+
+    async def test_state_isolation_between_tests(self):
```
Comment (jmelloy) [inline backend/tests/test_github_webhooks_phase2.py:456]: This test asserts `len(wh._debounce_queue._buffers.get('g-cl:t-cl1', [])) == 2` after external cancellation, but `_fire`'s CancelledError handler calls `self._buffers.pop(key, [])` when the task is still registered — which would clear the buffer to empty. Either the test is wrong about the expected behavior, or the `_fire` implementation is wrong. These two need to be reconciled.

File: backend/routes/webhooks.py [hunk 8]
```diff
@@ -48,6 +57,128 @@
...
+            name=f"foreman.github-debounced:{key}",
+        )
+        reset_foreman_poll(guild_id)
+        logger.info(
+            "github webhook debounce fired key=%s events=%d guild=%s",
+            key,
+            len(items),
+            guild_id,
+        )
+
```
Comment (jmelloy) [inline backend/routes/webhooks.py:117]: Calling `existing.cancel()` without awaiting means the cancelled task may still be running when the new `_fire` task starts. If both tasks overlap briefly, there is a window where `self._tasks[key]` points to the new task but the old task hasn't yet received its CancelledError, potentially leading to double delivery if the old task's sleep happens to complete before cancellation propagates.

File: backend/routes/webhooks.py [hunk 9]
```diff
@@ -48,6 +57,128 @@
...
+    def _deliver(self, key: str, guild_id: str, items: list[tuple[str, str | None]]) -> None:
+        summaries = [s for s, _ in items]
+        combined = "\n\n---\n\n".join(summaries) if len(summaries) > 1 else summaries[0]
+        # Use the last event's user_id: in typical bursts the final event is a
+        # human action (review, comment) that arrives after the CI bot events,
+        # so this attribution is usually more meaningful than the first event's.
+        user_id = items[-1][1]
+        spawn(
+            run_foreman_ai(guild_id, combined, user_id=user_id),
```
Comment (jmelloy) [inline backend/routes/webhooks.py]: Race condition: `schedule()` does `self._tasks.pop(key)` then immediately sets `self._tasks[key] = new_task` synchronously, so when the cancelled task's `CancelledError` handler runs, `self._tasks.get(key)` already points to the replacement task — not `current_task()`. This means the condition `self._tasks.get(key) is asyncio.current_task()` is always False for a normally-cancelled task, so the buffer-preservation branch is never reached and the buffer will be silently dropped on cancellation. Consider using a generation counter or a different sentinel to distinguish external vs. replacement cancellation.

File: backend/routes/webhooks.py [hunk 10]
```diff
@@ -35,6 +37,13 @@
 
 router = APIRouter()
 
+# How long (seconds) to wait for further events on the same PR before
+# delivering the buffered batch to the foreman.  GitHub often sends several
+# check_run completions and a review comment within a second of each other;
+# this window lets them coalesce into a single foreman invocation.
```
Comment (jmelloy) [inline backend/routes/webhooks.py:43]: The PR description says the debounce window is 3 seconds, but the default here is 60 seconds and `.env.example` also documents 60 seconds. A 60-second silence window means the foreman won't be invoked for at least a minute after any webhook activity, which seems unacceptably slow for a CI feedback loop. Please reconcile the intended default and document it clearly.

File: backend/tests/test_github_webhooks_phase2.py [hunk 11]
```diff
@@ -349,15 +348,142 @@ def test_webhook_dispatches_for_ci_bot_check_run(client):
...
+
+    @contextmanager
+    def _patched_env(self, wh):
+        """Context manager: fresh DebounceQueue instance + captured foreman calls.
+
+        Each call creates a brand-new DebounceQueue so tests cannot bleed state
+        into or out of the module-level singleton.
+        """
+        self._foreman_calls: list[tuple[str, str, str | None]] = []
+
```
Comment (jmelloy) [inline backend/tests/test_github_webhooks_phase2.py:380]: `patch.object(wh, 'spawn', side_effect=lambda coro, **kw: asyncio.create_task(coro))` replaces the production `spawn` with bare `asyncio.create_task`, which has different error-handling semantics (unhandled exceptions become unraisable errors rather than being logged). Tests calling `_deliver` indirectly through the real `_fire` path may emit noisy asyncio warnings or hide real bugs. Consider also patching `run_foreman_ai` at the `spawn` call site or asserting on task exception handling.

File: backend/tests/test_github_webhooks_phase2.py [hunk 12]
```diff
@@ -349,15 +348,142 @@ def test_webhook_dispatches_for_ci_bot_check_run(client):
...
+            wh._debounce_queue.schedule("g-ind:t-pr2", "g-ind", "pr2 event", "u4")
+            await asyncio.sleep(0.2)
+
+        assert len(self._foreman_calls) == 2
+        summaries = {c[1] for c in self._foreman_calls}
+        assert any("pr1 event" in s for s in summaries)
+        assert any("pr2 event" in s for s in summaries)
+
+    async def test_state_isolation_between_tests(self):
+        """Each _patched_env provides a fresh DebounceQueue with no state from prior tests."""
```
Comment (jmelloy) [inline backend/tests/test_github_webhooks_phase2.py:457]: This test asserts `len(wh._debounce_queue._buffers.get('g-cl:t-cl1', [])) == 2` after external cancellation, but the `_fire` CancelledError branch pops the buffer when `self._tasks.get(key) is asyncio.current_task()`. Whether this is True or False depends on the exact race between `schedule()`'s synchronous pop/assign and the async CancelledError injection — the test outcome is timing-dependent and may pass or fail non-deterministically.

**Review by jmelloy — COMMENTED** [review]
- The debounce implementation is well-structured with a clean `DebounceQueue` class, proper shutdown handling, and good test coverage across coalescing, separation, and isolation scenarios.
- **Critical race condition**: The `_fire` coroutine's CancelledError handler checks `self._tasks.get(key) is asyncio.current_task()`, but `schedule()` calls `self._tasks.pop(key)` and then assigns a new task — there is a window where `self._tasks[key]` is the replacement, meaning the cancelled task sees itself as no longer registered and incorrectly delivers the buffer immediately, causing a double-delivery.
- The default debounce window of **60 seconds** is far too large for a responsive CI/CD system; the PR description mentions 3 seconds but `.env.example` documents 60 seconds as the default — this mismatch could cause significant latency in foreman dispatch.
- `test_cancelled_events_not_lost` asserts that buffers are preserved after external cancellation, but the `_fire` CancelledError branch actually calls `self._buffers.pop(key, [])` when the task is still registered as current — so the buffer IS consumed and the test's assertion (`len(...) == 2`) will fail if the timing aligns such that the task is still the current registered one.
- The `_patched_env` context manager imports `routes.webhooks as wh` outside its scope and patches `wh.spawn` with `asyncio.create_task`, which bypasses any error handling or naming logic in the real `spawn` helper — tests may not accurately reflect production behavior for error cases.

File: backend/routes/webhooks.py (hunk 9 above)
Comment (jmelloy) [inline backend/routes/webhooks.py]: Race condition in cancellation guard: `_schedule_debounced_foreman` (i.e., `schedule`) pops and replaces `self._tasks[key]` before `cancel()` is awaited, so by the time `CancelledError` is injected into `_fire`, `self._tasks.get(key)` will already point to the *new* task, not `current_task()`. This means the guard `self._tasks.get(key) is asyncio.current_task()` will always be `False` for a normally-cancelled (replaced) timer, causing the buffer to never be delivered via the cancellation path — which is the intended behaviour. However this also means that if someone externally cancels the task *after* replacing it (an unlikely but possible race), events are silently dropped. The comment above should be clarified to reflect that the pop-before-cancel ordering is load-bearing.

File: backend/routes/webhooks.py (hunk 10 above)
Comment (jmelloy) [inline backend/routes/webhooks.py:43]: The default of 60 seconds contradicts the PR description which states '3-second window'. A 60-second debounce will introduce unacceptable latency for all webhook-driven automation. This should be 3 (matching the PR description) or at minimum the discrepancy should be explicitly justified in the comment and `.env.example`.

File: backend/routes/webhooks.py [hunk 13]
```diff
@@ -48,6 +57,128 @@
...
+        logger.info(
+            "github webhook debounce fired key=%s events=%d guild=%s",
+            key,
+            len(items),
+            guild_id,
+        )
+
+    async def _fire(self, key: str, guild_id: str) -> None:
+        """Sleep for the debounce window then deliver all buffered events.
+
```
Comment (jmelloy) [inline backend/routes/webhooks.py:120]: The `existing.cancel()` call in `schedule()` is not awaited — this is correct for fire-and-forget cancellation in sync code, but the CancelledError will only be injected into `_fire` on the next event-loop iteration. If `schedule()` is called again synchronously before yielding to the loop (e.g. in a tight loop), multiple cancelled tasks may exist transiently. This is unlikely to cause bugs given normal usage but is worth a comment.

File: backend/tests/test_github_webhooks_phase2.py [hunk 14]
```diff
@@ -349,15 +348,142 @@ def test_webhook_dispatches_for_ci_bot_check_run(client):
...
+        queue = wh.DebounceQueue(window_seconds=0.05)
+        with (
+            patch.object(wh, "_debounce_queue", queue),
+            patch.object(wh, "run_foreman_ai", new=fake_run_foreman),
+            patch.object(wh, "reset_foreman_poll"),
+            patch.object(wh, "spawn", side_effect=lambda coro, **kw: asyncio.create_task(coro)),
+        ):
+            yield
+
+    async def test_rapid_events_single_foreman_call(self):
```
Comment (jmelloy) [inline backend/tests/test_github_webhooks_phase2.py:396]: The `test_cancelled_events_not_lost` test asserts `len(wh._debounce_queue._buffers.get('g-cl:t-cl1', [])) == 2` after external cancellation. But per `_fire`'s `CancelledError` handler: if the task is still registered (which it is after external cancel with no replacement), it pops the buffer and calls `_deliver`. After `await asyncio.sleep(0.0)` the buffer will be empty and delivered — making the assertion incorrect and the subsequent `schedule` call only seeing the third event. This test likely passes only if the `CancelledError` handler path is not exercised as expected.

**Review by jmelloy — COMMENTED** [review]
- The debounce implementation is well-structured with a clean `DebounceQueue` class and good separation of concerns, but the `_fire` cancellation logic has a subtle race condition that could cause events to be silently dropped or the external-cancellation guard to behave unexpectedly.
- The default debounce window of 60 seconds is surprisingly large for a 'rapid coalescing' use case and will introduce significant latency for normal webhook events; the PR description mentions 3 seconds but the default is 60.
- The `test_cancelled_events_not_lost` test asserts that buffered events are preserved after external cancellation, but the `_fire` implementation actually pops and delivers (or discards) the buffer on `CancelledError` when it is still the registered task, meaning the assertion `len(wh._debounce_queue._buffers.get(...)) == 2` may be fragile or incorrect depending on timing.
- Tests in `TestDebounce` lack `@pytest.mark.asyncio` decorators (or equivalent configuration), which may cause them to run synchronously and pass trivially or fail depending on the pytest-asyncio mode configured.
- The `user_id` attribution logic (using last event's user_id) is reasonable but undocumented at the call site; the comment explaining this is only in `_deliver`, making it easy for future maintainers to miss the intentional design choice.

File: backend/tests/test_github_webhooks_phase2.py [hunk 15]
```diff
@@ -349,15 +349,312 @@ def test_webhook_dispatches_for_ci_bot_check_run(client):
...
+# ---------------------------------------------------------------------------
+# Debounce behaviour
+# ---------------------------------------------------------------------------
+
+
+class TestDebounce:
+    """Per-PR debounce timer: rapid events coalesce; separated events deliver separately."""
+
+    @asynccontextmanager
+    async def _patched_env(self, wh):
```
Comment (jmelloy) [inline backend/tests/test_github_webhooks_phase2.py:372]: The async test methods in `TestDebounce` have no `@pytest.mark.asyncio` decorator. Depending on the project's `asyncio_mode` pytest-asyncio setting, these may be silently collected but not actually awaited/run, giving false-positive passes. Confirm the project uses `asyncio_mode = 'auto'` in pytest config, or add explicit markers.

File: backend/routes/webhooks.py [hunk 16]
```diff
@@ -48,6 +57,140 @@
...
+        """Cancel all in-flight timers and wait for them to finish.
+
+        Awaiting the cancelled tasks lets them complete cleanly and prevents
+        "Task destroyed but pending" warnings.  Call this at server shutdown
+        and in async test teardown.
+        """
+        pending = [t for t in self._tasks.values() if not t.done()]
+        for task in pending:
+            task.cancel()
+        if pending:
```
Comment (jmelloy) [inline backend/routes/webhooks.py:96]: When an external caller cancels a task (e.g., test teardown or OS signal), `_tasks[key]` is left holding the cancelled/done task and is never removed. Subsequent `schedule()` calls will pop and discard it correctly, but `shutdown()` will redundantly attempt to cancel it. Consider also cleaning up `_tasks` inside `_fire` on `CancelledError` to keep state consistent.

File: backend/routes/webhooks.py [hunk 17]
```diff
@@ -48,6 +57,140 @@
...
+            await asyncio.gather(*pending, return_exceptions=True)
+        self._tasks.clear()
+        self._buffers.clear()
+        self._generation.clear()
+
+    def _deliver(self, key: str, guild_id: str, items: list[tuple[str, str | None]]) -> None:
+        summaries = [s for s, _ in items]
+        combined = "\n\n---\n\n".join(summaries) if len(summaries) > 1 else summaries[0]
```
Comment (jmelloy) [inline backend/routes/webhooks.py:104]: `_deliver` calls `spawn(run_foreman_ai(...))` which creates yet another asyncio task inside the already-async `_fire` task. This means errors in `run_foreman_ai` surface only in the spawned task's exception handler and are completely invisible to `_fire`. Since `_fire` is already async, consider awaiting `run_foreman_ai` directly here, or at minimum ensure the spawned task's exceptions are logged.

File: backend/routes/webhooks.py [hunk 18]
```diff
@@ -48,6 +57,140 @@
 _BOT_OK_EVENTS = frozenset({"check_run", "check_suite", "status"})
 
 
+class DebounceQueue:
+    """Per-PR debounce queue that coalesces rapid webhook events.
+
+    State is scoped to the instance so tests can create an isolated copy
+    without touching the module-level singleton.
```
Comment (jmelloy) [inline backend/routes/webhooks.py:64]: The `_tasks` type annotation `dict[str, asyncio.Task]` with `# type: ignore[type-arg]` suppresses the generic parameter. Python 3.9+ supports `dict[str, asyncio.Task[None]]` without ignore; using the proper type would remove the suppression and improve clarity.

**Review by jmelloy — COMMENTED** [review]
- The debounce implementation is well-structured with a clean `DebounceQueue` class, generation-counter race protection, and thorough test coverage including edge cases like stale generation and external cancellation.
- The `_deliver` method uses `spawn()` to wrap `run_foreman_ai`, but since `spawn` is patched in tests to use `asyncio.create_task`, the actual delivery path is correct; however `_fire` is itself an async task so `_deliver` calling `spawn` inside an async context adds an extra task layer that may be unnecessary.
- The `test_cancelled_events_not_lost` test documents that externally-cancelled timers preserve the buffer, but the `_tasks` dict is NOT cleaned up when a task is externally cancelled — `_tasks[key]` still holds the done/cancelled task, which means `schedule()` will pop it correctly but `shutdown()` will attempt to cancel an already-done task (harmless but slightly wasteful).
- The user_id attribution strategy (using the last event's user_id) is noted in a comment but could silently change attribution if a CI bot event trails a human event, which may cause unexpected behavior in downstream foreman logic.
- Test class `TestDebounce` lacks `@pytest.mark.asyncio` or equivalent markers on individual async test methods — depending on the pytest-asyncio mode configured, these tests may be silently skipped rather than run.

**Comment by jmelloy** [comment]
Superseded by #401, which addresses all review comments (generation-counter race fix, asyncio.create_task + error callback, bot-user attribution fix, buffer snapshot on cancel). The missing shutdown_debouncer hook from this PR has now been added to #401 (commit 69c553f).
